File: src/tatami.pl
use feature ':5.10';

sub Tatami
{
  my ($arg1) = @_;
  use constant nMax => 100000000;
  use constant nMaxSqrt => int(sqrt(nMax));
  foreach my $i (7..nMaxSqrt - 1) {
    $k2 = $i + 3;
    $k3 = $i + $i - 4;
    foreach my $k ($k2..$k3) {
      last if (($i * $k2) >= nMax);
      $k4 = int(nMax / $i);
      if ($k3 < $k4) {
        $k4 = $k3;
      }
      if (($i & 1) == 1) {
        if (($k2 & 1) == 0) {
          for ($j = $k2; $j <= $k4; $j += 2) {
             $v[$i * $j]++;
          }
        }
      }
      else {
        foreach my $j ($k2..$k4) {
          $v[$i * $j]++;
        }
      }
      $k2 += $i + 1;
      $k3 += $i - 1;
    }
  }
  foreach my $i (0..nMax - 1) {
    if ($v[$i] == $arg1) {
      return $i;
    }
  }
}

say "The smallest room size s for which T(s) = 200 is ", Tatami(200);

File: src/tatami.c
// Tatami.c

#include <math.h>
#include <stdio.h>

#define nMax 100000000

unsigned char v[nMax];

int Tatami(int s)
{
    const int nMaxSqrt = sqrt(nMax);

    for (int i = 0; i < nMaxSqrt; i++)
    {
        int k2, k3;
        for (int k = 1;
            (k2 = ((k * i) + k + 2)) <= (k3 = (((k + 1) * i) - k - 3));
            ++k)
        {
            if ((i * k2) >= nMax)
                break;
            for (int j = k2; j <= k3; ++j)
            {
                if ((i & j & 1) || (i * j >= nMax))
                    continue;
                ++v[i * j];
            }
        }
    }

    for (int i = 0; i < nMax; ++i)
        if (v[i] == s)
            return i;
}

int main()
{
    printf("The smallest room size s for which T(s) = 200 is %d\n", Tatami(200));
}
